examples/linear_model/plot_tweedie_regression_insurance_claims.py

```diff
@@ -429,40 +435,54 @@ def score_estimator(
     df_test,
     target="ClaimAmount",
     weights="Exposure",
+    power=2-eps,
 )
 print(scores)
 
 
 ##############################################################################
 #
-# Indeed, an alternative approach for modeling the total loss is with a unique
-# Compound Poisson Gamma model, also corresponding to a Tweedie model with a
-# power :math:`p \in (1, 2)`. We determine the optimal hyperparameter ``p``
-# with a grid search:
+# Instead of taking the product of two independently fit models for frequency
+# and severity one can directly model the total loss is with a unique Compound
+# Poisson Gamma generalized linear model (with a log link function). This
+# model is a special case of the Tweedie model with a power parameter :math:`p
+# \in (1, 2)`.
+#
+# We determine the optimal hyperparameter ``p`` with a grid search so as to
+# minimize the deviance:
 
 from sklearn.model_selection import GridSearchCV
 
-# exclude upper bound as power>=2 does not support y=0.
-params = {"power": np.linspace(1 + 1e-4, 2 - 1e-4, 8)}
-
-
-# this takes a while
-glm_total = GridSearchCV(
-    TweedieRegressor(tol=1e-3, max_iter=500), cv=3,
-    param_grid=params, n_jobs=-1
+# exclude upper bound as power>=2 as p=2 would lead to an undefined unit
+# deviance on data points with y=0.
+params = {"power": np.linspace(1 + eps, 2 - eps, 5)}
+
+X_train_small, _, df_train_small, _ = train_test_split(
+    X_train, df_train, train_size=5000)
+
+# This can takes a while on the full training set, therefore we do the
+# hyper-parameter search on a random subset, hoping that the best value of
+# power does not depend too much on the dataset size. We use a bit
+# penalization to avoid numerical issues with colinear features and speed-up
+# convergence.
+glm_total = TweedieRegressor(max_iter=10000, alpha=1e-2)
+search = GridSearchCV(
+    glm_total, cv=3,
+    param_grid=params, n_jobs=-1, verbose=10,
+    refit=False,
```

Comment:
This gridsearch is still using r2 for its scoring, but the description says we are using grid search to minimize deviance.

There may not be a nice way to do this, because we are searching over power, the `mean_tweedie_deviance` metric depends on power.

Reply:
Grid search is maximizing D2 (explained deviance ratio) which is the same as minimizing deviance which itself depends on power...

Reply:
I am confused here: isn't `GridSearch` going to compare scores that are not comparable?

`est.score` depends on the power and the power is being searched over.

Reply:
@NicolasHug I share your concern. I am not sure if choosing the value of power that maximizes D^2 is a methodological error or not. @lorentzenchr @agramfort what is your opinion?

For reference here is the meaning of the deviance of a GLM:
- https://bookdown.org/egarpor/PM-UC3M/glm-deviance.html
- https://en.wikipedia.org/wiki/Deviance_(statistics)
- https://en.wikipedia.org/wiki/Tweedie_distribution#The_Tweedie_deviance

Reply:
If D2 is not a suitable metric for model selection, then we would need to choose another scoring method that would make more sense. I don't think MAE is a good model selection for positive regression with long tailed distributions in y. MAE makes most sense when the y values stay on the same scale.

Alternative one could add a register gini_coefficient_score which is defined in terms of ratios of areas under the [Lorenz curve](https://en.wikipedia.org/wiki/Lorenz_curve).

At the moment, I manually compute the cumulated claims curve and its area under curve in the last plot of the example:

![image](https://user-images.githubusercontent.com/89061/66548131-f2799780-eb40-11e9-909e-954cf3e2752c.png)

This is the same as the Lorenz curve with a swapped x axis (I used from riskiest to safest instead of from safest to riskiest for the Lorenz curve).

The problem with the Gini coefficient is that it is only a ranking metric (like ROC AUC is a ranking metric for binary classification): it stays invariant if you apply any monotonic transformation to y_pred. It therefore does not penalize badly calibrated models.

Reply:
On the question on how to choose the optimal p number it is a tricky point. @lorentzenchr discussed it in https://github.com/scikit-learn/scikit-learn/pull/9405#issuecomment-506884728
> For the metric, I think it should be specified in advance, based on prior knowledge, business goals, etc. It must be the same for all models being compared. As the deviance is differentiable in p, it would be interesting to see, when the model ranking changes.
Slightly related is this PR #11430: As the goal is to estimate the expected value, any Tweedie deviance is (theoretically) fine, as they all are striclty consistent for the expected value.
I don't have a better answer right now.

For instance, [Shono 2018](http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.637.6974&rep=rep1&type=pdf) seem to be doing a somewhat comparable approach,
> In this study, we carried out the parameter estimation based onthe following procedure.
>  (i) Estimate the power parameter (p) by maximizing the profilelog-likelihood across the grid values of (p) (see Figs. 2 and 6) in the range of 1<p<2 through the explicit form of the probability density function in Eq.(7).
>   (ii) Estimate the regression coefficients using the framework of quasi-likelihood in the GLM fixing the value of pin the estimate obtained in the step (i) also based on the formulae of Eq.(7)including the deviance.

A more detailed discussion can be found in [Dunn and Smyth 2002](https://link.springer.com/content/pdf/10.1007/s11222-005-4070-y.pdf)
> Estimation of p a more difficult problem than estimating β or φ.Most authors using Tweedie densities have taken p to be specified apriori. Jørgensen (1987), in analysing the amount of money spent by Amazonian peasants hiring outside labour, chooses p=1.75 and explicitly states the choice is somewhat arbitrary. Likewise, Nelder (1994) arbitrarily sets p=1.5 whenan alysing  the  time  spent  splicing  cables.  Given  the  ability  to compute the maximum likelihood estimator φ conditional on p,the maximum likelihood estimate of p and an approximate confidence interval can obtained by evaluating the profile likelihood for p on a grid of values followed by a univariate optimization. A similar approach was used by Gilchrist and Drinkwater (1999) and Smyth (1996) to estimate p.
> [..] The  most  commonly  used  estimators  forφin  generalized linear models are the mean deviance estimator

So using grid search over values of p (particularly in for the compound Poisson Gamma (1 < p < 2) might be OK, as far as I understand. In any case, I do find the deviance metric more intuitive (due to it being a generalization of MSE) than the Gini coefficient.

Reply:
> using grid search over values of p (particularly for the compound Poisson Gamma (1 < p < 2) might be OK

what about when `p` leads to completely different distributions? e.g. the unit deviance for normal is
` dev = (y - y_pred)**2`
vs poisson
`dev = 2 * (xlogy(y, y/y_pred) - y + y_pred)`

Are these really comparable? I didn't plot them but I wouldn't be surprised if the normal was consistently higher. The selected `p` would always come from the normal.

Reply:
Thanks @rth for the citations and @ogrisel for your many contributions lately. I'll try to further clarify my point of view:

- For model fitting, we could fully specify a stochastic model and obtain all parameters by maximum likelihood, i.e. not only ``coef_`` but also the power parameter ``p``. The deviance is a part of the log-likelihood. Unfortunately, it does not contain all dependency on ``p`` (but all that is needed for the expected value). Therefore, we should not use a grid search to find an optimal value for ``p`` based on deviance, that depends on this same ``p``. We should set it a priori. One could even fit two models, one with, say, p=1.5 and another with p=1.9.
- For model comparison with a deviance, I strongly recommend to set ``power`` in advance or use several metrics, say, one with p=1.5, another with p=1.9. Other metrics like MAE are not suitable here, because they elicit the median (or something else), but not the expected value (conditional on X), which is our goal.
- D2 is equivalent to mean_tweedie_deviance with the same ``power`` parameter when comparing models.
- Wanna use ``GridSearch``: One could specify a metric a priori, say ``power=1.5`` and use this metric to find the best value of ``p`` via CV. I expect to find ``power≈p``, but this may also depend on other hyperparameters, mostly the penalty.

main message: distinction between loss (training, ``p``) and metric (model comparison on test/validation data, ``power``)

Reply:
Thanks for you feedback @lorentzenchr. In #15176 I have updated the example to use the Gini index as a model selection metric in the grid search. It also selects p=2.

> Wanna use GridSearch: One could specify a metric a priori, say power=1.5 and use this metric to find the best value of p via CV. I expect to find power≈p, but this may also depend on other hyperparameters, mostly the penalty.

That makes sense, I think we should try to implement this in the example. I also expect the feature engineering to have a potential impact.

Reply:
> we should not use a grid search to find an optimal value for p based on deviance

So since `est.score` is the deviance, we should not allow users to grid search on `p`? I don't think we have an API for forbidding users to grid search on something

Reply:
No need for forbidding. Proper documentation and examples are enough in my opinion.

Reply:
I read a bit more about Tweedie estimators for Compound Poisson Gamma models, an indeed the correct way to select the model would be to fit a grid of GLMs for various power parameters in [1.1 - 1.9], for each of them estimate the corresponding \Phi parameter using the Tweedie variance formula that links \Phi, \mu and p together and the empirical variance of each fitted model on the training set, then compute the negative log likelihood on a held out set for each of them and select the best power that minimizes the NLL.

This is quite cumbersome to do with our current API. Instead I will remove the grid search for p from this example in this PR and compare the 2 models (freq * severity product model vs compound Poisson gamma single model with fixed power=1.7 parameter) for 3 values of Tweedie deviance power  (1.5, 1.7 and 1.9). In a quick experiment, the best model is the best for the 3 values.

Reply:
> Instead I will remove the grid search for p from this example in this PR and compare the 2 models (freq * severity product model vs compound Poisson gamma single model with fixed power=1.7 parameter) for 3 values of Tweedie deviance power (1.5, 1.7 and 1.9).

Sounds good to me. For future reference, I did quick benchmark to evaluate optimal power parameters we obtain for D2 metrics with different fixed `power` values in https://github.com/rth/notebooks/blob/master/sklearn/plot_tweedie_regression_insurance_claims_optimal_power.ipynb

The conclusion seem to be that,
 - there is quite a lot of noise between different CV folds, so I'm not sure to what extent we should rely these on results anyway.
 - results are indeed dependent on the regularization as mentioned by @lorentzenchr,
![alpha01](https://user-images.githubusercontent.com/630936/66751273-ef5b1000-ee8e-11e9-91bb-5eeffddbb03a.png)
![alpha1](https://user-images.githubusercontent.com/630936/66751288-f84be180-ee8e-11e9-9a77-6ccdb70c69cb.png)
![alpha10](https://user-images.githubusercontent.com/630936/66751294-fb46d200-ee8e-11e9-83e3-0b8766f29e4b.png)
 - in this particular case, the optimum seem to be reached in `power=2.` for a number of evaluation metrics (which I suppose is not the expected result). Here I increased CV folds to 5 and am using a grid of 10 points in `power` between (1, 2) as well as the additional point [0]. 

For the purpose of moving this PR forward I think simplifying that example without CV should be enough, but in general it's a challenging topic and maybe we should open a follow up issue on this.

Reply:
I have pushed the example simplification that removes the GridSearchCV but print deviances for various values of p.

I also made the two models predict the pure premium (total claim amount per unit of exposure) so that they are more directly comparable.

I also tweaked the regularization parameter a bit.

I also fixed the Lorenz curves plot: the exposure is only used for y. Which should fixed the conceptual issue I had in #15176. I will need to update that later PR and the other example accordingly.

Reply:
Thanks @rth for the D2 based benchmark. But what @lorentzenchr suggested is instead to pick a deviance metric with an arbitrary power a priory and grid search CV for the best model power hyperparam and see if they match.

Alternatively, using the complete log-likelihood as the model selection metric is describe in section 4.2 of this paper (in the context of Tweedie GBRT but the concept stays the same): https://arxiv.org/abs/1508.06378

or alternatively in this R package:

https://rdrr.io/cran/tweedie/man/tweedie.profile.html

this is called profile evaluation of the power parameter. However it's a bit complicated because it requires the additional estimation of the phi hyperparam and I am not sure it bring that much benefit in practice.